examples/test_construct_spec.rb

```diff
@@ -0,0 +1,116 @@
+# Run with:
+#   rspec --format doc --order defined examples/test_construct_spec.rb
+
+# We can't replace the usage of the instance variable with a let or method call,
+# since the example isn't available then, so we disable the warning:
+#   rubocop:disable RSpec/InstanceVariable
+
+require "test_construct/rspec_integration"
+
+RSpec.describe TestConstruct do
+  before do
+    stub_const("TEST_PATH", "alice/rabbithole")
+    stub_const("TEST_FILE", "white_rabbit.txt")
+    stub_const("TEST_CONTENTS", "I'm late!")
+  end
```

Comment:
I'm afraid I don't understand this code. What's the reason we are doing this instead of 

```
let(:test_path) { "alice/rabbithole" }
let(:test_file) { "white_rabbit.txt" }
let(:test_contents) { "I'm late!" }
```

I made that change locally and it seems to work, but I very well may be missing something.

Reply:
It will work either way. I think it is a question of documenting the intent and how `let` is intended to be used.

What we want is a constant that can be used in multiple tests. Normally, one would just write:
```ruby
RSpec.describe TestConstruct do
  TEST_PATH = "alice/rabbithole"
  TEST_FILE = "white_rabbit.txt"
  TEST_CONTENTS = "I'm late!"
...
```
Any programmer would immediately recognize these are constants. The problem with that is explained in the [rubocop-rspec documentation](https://rubocop-rspec.readthedocs.io/en/latest/cops_rspec/#rspecleakyconstantdeclaration):

> Constants, including classes and modules, when declared in a block scope, are defined in global namespace, and leak between examples.

See [RSpec issue #106](https://github.com/rspec/rspec-core/issues/106) for details.

Using `stub_const` in a `before` block is the solution [rubocop-rspec](https://github.com/rubocop-hq/rubocop-rspec/) recommends for the leaky constant problem. RSpec provides [`stub_const`](https://relishapp.com/rspec/rspec-mocks/v/3-9/docs/mutating-constants/stub-undefined-constant) to allow constants to be stubbed (whether they are defined in the code under test or not). I think the method name `stub_const` clearly conveys the intent (defining a constant) so using it makes sense to me.

In contrast, RSpec provides [let](https://relishapp.com/rspec/rspec-core/docs/helper-methods/let-and-let). The RSpec documentation says:

> Use let to define a memoized helper method. The value will be cached across
multiple calls in the same example but not across examples.

The [method documentation](https://www.rubydoc.info/github/rspec/rspec-core/RSpec%2FCore%2FMemoizedHelpers%2FClassMethods%3Alet) also includes these notes:

> Note: `let` can enhance readability when used sparingly (1,2, or maybe 3 declarations) in any given example group, but that can quickly degrade with overuse. YMMV.

> Note: Because `let` is designed to create state that is reset between each example, and `before(:context)` is designed to setup state that is shared across all examples in an example group, `let` is not intended to be used in a `before(:context)` hook.

> Generates a method whose return value is memoized after the first call. Useful for reducing duplication between examples that assign values to the same local variable.

We don't need a method created nor do we need the value memoized so using `let` for a constant seems inappropriate to me. The fact that `let` is not intended to be used in a `before(:context)` hook only makes in less desirable--I shouldn't have to worry about where I am using a constant (other than scope).

I generally try to follow ThoughBot's [_Best Practices_](https://github.com/thoughtbot/guides/tree/master/best-practices) unless I have a good reason not to and they recommend not using`let`.  See [_Let's Not_](https://thoughtbot.com/blog/lets-not) for why. As a result, I'm not generally a fan of using `let` anyway.

Reply:
Thanks for the detailed explanation. Makes sense to me!